### PR TITLE
ALSA/ASoC: SOF: Intel: Use PIO mode fro HDA commands instead of CORB on Lunar Lake

### PR DESCRIPTION
**Pull request by ujfalusi**
It is recommended to use Immediate Command Response (PIO mode) instead of CORB on Lunar Lake family for HDA commands.

This PR adds support for PIO mode commands while keeping RIRB functionality for receiving URs

### CODE REVIEW

**Review by lgirdwood — DISMISSED** [review]

File: include/sound/hdaudio.h [hunk 1]
```diff
@@ -390,6 +391,10 @@ int snd_hdac_bus_exec_verb_unlocked(struct hdac_bus *bus, unsigned int addr,
 void snd_hdac_codec_link_up(struct hdac_device *codec);
 void snd_hdac_codec_link_down(struct hdac_device *codec);
 
+int snd_hdac_bus_send_cmd_pio(struct hdac_bus *bus, unsigned int val);
+int snd_hdac_bus_get_response_pio(struct hdac_bus *bus, unsigned int addr,
+				  unsigned int *res);
+
```
Comment (plbossart) [inline include/sound/hdaudio.h]: The commit message is hard to parse due to missing punctuation.

The PIO (Immediate Command Response) can only be used when the CORB is
disabled, inactive but RIRB still can receive unsolicited responses.

Did you mean:
The PIO (Immediate Command Response) can only be used when the CORB is
disabled and hence inactive. RIRB can be enabled independently and can still be used to handle unsolicited responses.

File: sound/hda/hdac_controller.c [hunk 2]
```diff
@@ -135,6 +136,88 @@ static unsigned int azx_command_addr(u32 cmd)
 	return addr;
 }
 
+/* receive a response via Immediate Response */
```
Comment (plbossart) [inline sound/hda/hdac_controller.c]: receive an Immediate Response with PIO?

File: sound/soc/sof/intel/hda-bus.c [hunk 3]
```diff
@@ -72,7 +78,14 @@ void sof_hda_bus_init(struct snd_sof_dev *sdev, struct device *dev)
...
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
-	snd_hdac_ext_bus_init(bus, dev, &bus_core_ops, sof_hda_ext_ops);
+	const struct sof_intel_dsp_desc *chip = get_chip_info(sdev->pdata);
+
+	if (chip && chip->hw_ip_version < SOF_INTEL_ACE_2_0) {
+		snd_hdac_ext_bus_init(bus, dev, &bus_core_ops, sof_hda_ext_ops);
+	} else {
+		snd_hdac_ext_bus_init(bus, dev, &bus_core_pio_ops, sof_hda_ext_ops);
+		bus->use_pio_for_commands = true;
+	}
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-bus.c]: Don't you need a LunarLake-specific quirk for snd-hda-intel as well? CORB/PIO are not SOF-specific.

Comment (plbossart) [reply sound/soc/sof/intel/hda-bus.c]: I would actually add a new bit in AZX capabilities and do everything in sound/hda. There's no reason to do anything in SOF regarding commands.

e.g. add something in hda_controller.h 
````
#define  AZX_DCAPS_PIO_COMMANDS 31
````
and add the quirk in the PCI data.

File: sound/hda/hdac_controller.c [hunk 4]
```diff
@@ -135,6 +136,88 @@ static unsigned int azx_command_addr(u32 cmd)
...
+static int snd_hdac_bus_wait_for_pio_response(struct hdac_bus *bus,
```
Comment (ranj063) [inline sound/hda/hdac_controller.c:140]: @ujfalusi doesnt it make sense to only add the bool in this patch and add the helpers for PIO helpers in the next patch where they're actually used?

File: include/sound/hdaudio.h (hunk 1 above)
Comment (ujfalusi) [reply include/sound/hdaudio.h]: @plbossart, PIO actually can be used when CORB is enabled but inactive (no commands pending), hence the strange wording.
RIRB on the other hand can be kept enabled. The response will come back to the interface from where the command was sent (IC or CORB).

File: sound/hda/hdac_controller.c (hunk 4 above)
Comment (ujfalusi) [reply sound/hda/hdac_controller.c:140]: But the bool alone will not do anything, we need the functions which can be used to send the commands via PIO and at the same time to keep the CORB disabled.

File: sound/soc/sof/intel/hda-bus.c (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-bus.c]: > Don't you need a LunarLake-specific quirk for snd-hda-intel as well? CORB/PIO are not SOF-specific.

Yes, we need a separate quirk for legacy HDA, or a way to use PIO command in case of Lunar Lake.
This PR is not complete, it is for SOF so we can run complete validation test on it.

Comment (ujfalusi) [reply sound/soc/sof/intel/hda-bus.c]: > I would actually add a new bit in AZX capabilities and do everything in sound/hda. There's no reason to do anything in SOF regarding commands.

The two stack is independent in this regards, afaik.
We need separate ways to set different `hdac_bus_ops` via legacy and SOF stack _and_ to set the new `use_pio_for_commands` flag at the same time.

> e.g. add something in hda_controller.h
> 
> ```
> #define  AZX_DCAPS_PIO_COMMANDS 31
> ```

The legacy stack do not support this type of operation (hda_controller w/ azx_*), it can fallback to pure PIO mode w/o RIRB and thus no URs, so the legacy is a bit more involved.

> 
> and add the quirk in the PCI data.

Sure, but the legacy (using hda_controller) and SOF (using hdac_controller) do not share code that we could do this in one place... afaik, imho.

**Comment by ujfalusi** [comment]
Changes since v1:
- separate the flag introduction
- Add quirk flag for legacy to use: AZX_DCAPS_PIO_COMMANDS
- do not export the new PIO send_cmd/get_response and select between CORB/RIRB and PIO in core code based on use_pio_for_commands flag
- enable the PIO mode for LNL in legacy and SOF stack.
- dropped the SOF patch to use snd_hdac_bus_exec_verb_unlocked() as the PIO vs CORB selection is done in core now.

**Comment by ujfalusi** [comment]
Changes since v2:
- fix prototype comment for snd_hdac_bus_get_response_rirb()

File: sound/pci/hda/hda_intel.c [hunk 5]
```diff
@@ -2503,7 +2503,8 @@ static const struct pci_device_id azx_ids[] = {
 	{ PCI_DEVICE_DATA(INTEL, HDA_RPL_PX, AZX_DRIVER_SKL | AZX_DCAPS_INTEL_SKYLAKE) },
 	{ PCI_DEVICE_DATA(INTEL, HDA_MTL, AZX_DRIVER_SKL | AZX_DCAPS_INTEL_SKYLAKE) },
 	/* Lunarlake-P */
-	{ PCI_DEVICE_DATA(INTEL, HDA_LNL_P, AZX_DRIVER_SKL | AZX_DCAPS_INTEL_SKYLAKE) },
+	{ PCI_DEVICE_DATA(INTEL, HDA_LNL_P, AZX_DRIVER_SKL | AZX_DCAPS_INTEL_SKYLAKE |
+					    AZX_DCAPS_PIO_COMMANDS) },
```
Comment (plbossart) [inline sound/pci/hda/hda_intel.c]: we could also introduce a macro such as  AZX_DRIVER_LNL where all the defaults are set, that will save lines for future additions.

File: sound/soc/sof/intel/hda-bus.c [hunk 6]
```diff
@@ -72,7 +72,12 @@ void sof_hda_bus_init(struct snd_sof_dev *sdev, struct device *dev)
 
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_LINK)
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
+	const struct sof_intel_dsp_desc *chip = get_chip_info(sdev->pdata);
+
 	snd_hdac_ext_bus_init(bus, dev, &bus_core_ops, sof_hda_ext_ops);
+
+	if (chip && chip->hw_ip_version == SOF_INTEL_ACE_2_0)
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-bus.c:79]: "> SOF_INTEL_ACE_1_0" would be more future-proof

File: sound/pci/hda/hda_controller.c [hunk 7]
```diff
@@ -1180,6 +1180,9 @@ int azx_bus_init(struct azx *chip, const char *model)
 	if (chip->driver_caps & AZX_DCAPS_4K_BDLE_BOUNDARY)
 		bus->core.align_bdle_4k = true;
 
+	if (chip->driver_caps & AZX_DCAPS_PIO_COMMANDS)
+		bus->core.use_pio_for_commands = true;
+
```
Comment (plbossart) [inline sound/pci/hda/hda_controller.c:1185]: I would split the library from the sound/pci/hda/hda_controller.c in two separate patches.

Arguably the sound/pci/hda/hda_controller.c stuff should be combined with the next patch, it would make it easier to bisect regressions or changes in behavior (firmware download, etc).

How many engineers does it take to split code, eh?

**Review by plbossart — COMMENTED** [review]
I only have cosmetic comments @ujfalusi this looks pretty good to me!

File: sound/pci/hda/hda_controller.c (hunk 7 above)
Comment (ujfalusi) [reply sound/pci/hda/hda_controller.c:1185]: The problem is that the only place where the CORB DMA is enabled is via the `snd_hdac_bus_init_cmd_io()` in hdac_controller.c and that needs to be done at the same time as we allow hda_controller to use PIO commands.

Hrm, unless I split the patch and have the hdac_controller.c changes in the first one which anyways also needs this.

Comment (ujfalusi) [reply sound/pci/hda/hda_controller.c:1185]: I will still keep the enabling this as separate patches.

File: sound/pci/hda/hda_intel.c (hunk 5 above)
Comment (ujfalusi) [reply sound/pci/hda/hda_intel.c]: Right, my optimistic thinking was that this is only going to be needed for LNL, but ACE2.0 variants will need this as well.
OK.

File: sound/soc/sof/intel/hda-bus.c (hunk 6 above)
Comment (ujfalusi) [reply sound/soc/sof/intel/hda-bus.c:79]: I hope that  `> ACE_2_0` will not need this, if it does then we change the condition.

File: sound/hda/hdac_controller.c (hunk 2 above)
Comment (ujfalusi) [reply sound/hda/hdac_controller.c]: ok

**Comment by ujfalusi** [comment]
Changes since v3:
- split the hdac_controller/hda_controller patch which implements support for use_pio_for_commands
- Create AZX_DCAPS_INTEL_LNL for the quirk definition
- taken out from draft

**Review by plbossart — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Comment by plbossart** [comment]
@ujfalusi should we merge this?

**Comment by ujfalusi** [comment]
> @ujfalusi should we merge this?

I think we should and send it for 6.10 as early as possible to get feedback on the legacy HDA change.

**Review by lgirdwood — APPROVED** [review]

**Comment by marc-hb** [comment]
Revert just submitted in:
- #4892

... to avoid these mass test failures:
- #4889
